[Train] [doc] Local mode user guide (#57751)
ray-project/ray: doc/source/train/user-guides/local_mode.rst

.. _train-local-mode:

> Local Mode

Reviewer: ```suggestion
Local mode
```

==========

.. important::

>     This user guide shows how to use local mode for Ray Train V2 only.

Reviewer 2: Link to https://github.com/ray-project/ray/issues/49454, mention that v2 is enabled by default starting from ray version X, and remove the bolded line

    **This user guide assumes that you're using Ray Train V2 with the environment variable ``RAY_TRAIN_V2_ENABLED=1`` enabled.**

Local mode in Ray Train allows you to run your training function directly in the local process.

> This is useful for testing, development, and quick iterations.

Reviewer 2: ```suggestion
This is useful for development, testing, and rapid iteration.
```

When to use local mode
----------------------

> Local mode is particularly useful in the following scenarios:

Reviewer 2: I think it's worth highlighting that users can easily switch between Ray Train and local mode for debugging purposes with minimal code changes. Because every `ray.train` function has a local analogue, all users need to do is set `num_workers=0` and the code should still run without any changes.

            if ray.train.get_context().get_world_size() > 1:
                train_loader.sampler.set_epoch(epoch)

            epoch_loss = 0.0
            for batch_idx, (images, labels) in enumerate(train_loader):
                outputs = model(images)
                loss = criterion(outputs, labels)
                
                optimizer.zero_grad()
                loss.backward()
                optimizer.step()
                
                epoch_loss += loss.item()

            avg_loss = epoch_loss / len(train_loader)

            # Report metrics and checkpoint
            with tempfile.TemporaryDirectory() as temp_dir:
                torch.save(model.state_dict(), os.path.join(temp_dir, "model.pt"))
                ray.train.report(
                    {"loss": avg_loss, "epoch": epoch},
                    checkpoint=Checkpoint.from_directory(temp_dir)
                )

    # Configure trainer for local mode
    trainer = TorchTrainer(
        train_loop_per_worker=train_func,
        train_loop_config={"lr": 0.001, "epochs": 10, "batch_size": 32},
        scaling_config=ScalingConfig(num_workers=0, use_gpu=True),
    )
    result = trainer.fit()

> Then, launch the training with torchrun:

Reviewer 2: ```suggestion
Then, launch training with torchrun:
```

.. code-block:: bash

    # Train on 4 GPUs on a single node
    RAY_TRAIN_V2_ENABLED=1 torchrun --nproc-per-node=4 train_script.py

Ray Train automatically detects the torchrun environment variables and configures the distributed
training accordingly. You can access distributed training information through :func:`ray.train.get_context()`:

> .. testcode::

Author: I added one section at the end to list the different behaviors.

    print(f"Local rank: {context.get_local_rank()}")

Multi-node multi-GPU training
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

You can also use torchrun to launch multi-node training with local mode. The following example shows
how to launch training across 2 nodes with 4 GPUs each:

On the master node (``192.168.1.1``):

.. code-block:: bash

    RAY_TRAIN_V2_ENABLED=1 torchrun \
        --nnodes=2 \
        --nproc-per-node=4 \
        --node_rank=0 \
        --rdzv_backend=c10d \
        --rdzv_endpoint=192.168.1.1:29500 \
        --rdzv_id=job_id \
        train_script.py

On the worker node:

.. code-block:: bash

    RAY_TRAIN_V2_ENABLED=1 torchrun \
        --nnodes=2 \
        --nproc-per-node=4 \
        --node_rank=1 \
        --rdzv_backend=c10d \
        --rdzv_endpoint=192.168.1.1:29500 \
        --rdzv_id=job_id \
        train_script.py

.. note::
    When using torchrun with local mode, Ray Data isn't supported. The training relies on standard
    PyTorch data loading mechanisms.

Using local mode with Ray Data
-------------------------------

Single-process local mode works seamlessly with Ray Data for data loading and preprocessing.

> The data is processed and provided to your training function without distributed workers.

Reviewer 2: ```suggestion
Ray Data processes your data and provides it to your training function without distributed workers.
```

Testing with local mode
-----------------------

Local mode is excellent for unit testing your training logic. The following example shows how to write
a unit test with local mode:

.. testcode::
    :skipif: True

    import unittest
    import ray
    from ray.train import ScalingConfig
    from ray.train.torch import TorchTrainer

    class TestTraining(unittest.TestCase):
        def test_training_runs(self):
            def train_func(config):
                # Minimal training logic
                ray.train.report({"loss": 0.5})

            trainer = TorchTrainer(
                train_loop_per_worker=train_func,
                scaling_config=ScalingConfig(num_workers=0),
            )
            result = trainer.fit()

            self.assertIsNone(result.error)
            self.assertEqual(result.metrics["loss"], 0.5)

    if __name__ == "__main__":
        unittest.main()

Limitations
-----------

Local mode has the following limitations:

* **No Ray Train fault tolerance**: Worker-level fault tolerance doesn't apply in local mode.
* **Ray Data limitations**: Ray Data isn't supported when using torchrun with local mode for multi-process training.

> * **Different behavior**: Some framework-specific behaviors might differ slightly from Ray Train's distributed mode.

Reviewer 2: Maybe give an example?

The following example shows how to use Ray Data with single-process local mode:

.. testcode::
    :skipif: True

    import ray
    from ray.train import ScalingConfig
    from ray.train.torch import TorchTrainer

    def train_func(config):
        # Get the dataset shard
        train_dataset = ray.train.get_dataset_shard("train")

        # Iterate over batches
        for batch in train_dataset.iter_batches(batch_size=32):
            # Training logic
            pass

    # Create a Ray Dataset
    dataset = ray.data.read_csv("s3://bucket/data.csv")

    trainer = TorchTrainer(
        train_loop_per_worker=train_func,
        scaling_config=ScalingConfig(num_workers=0),
        datasets={"train": dataset},
    )
    result = trainer.fit()

.. note::
    Ray Data isn't supported when using torchrun with local mode for multi-process training.

Testing with local mode
-----------------------

Local mode is excellent for unit testing your training logic. The following example shows how to write
a unit test with local mode:

.. testcode::
    :skipif: True

    import unittest
    import ray
    from ray.train import ScalingConfig
    from ray.train.torch import TorchTrainer

>     class TestTraining(unittest.TestCase):

Reviewer 3: Can you use `pytest` instead of `unittest` here?

API behavior differences in local mode
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

The following ``ray.train`` APIs have different behavior in local mode:

* :func:`ray.train.report`:

  * Checkpoints aren't persisted to storage. They're only stored in memory.
  * The ``checkpoint_upload_mode``, ``checkpoint_upload_fn``, ``validate_fn``, and ``delete_local_checkpoint_after_upload`` parameters have no effect.
  * Metrics are logged locally instead of being sent through the reporting pipeline.
  * No synchronization barrier across workers.

* :func:`ray.train.get_checkpoint`:

  * Returns the last checkpoint stored in memory from the current run.
  * Doesn't load checkpoints from persistent storage.

* :func:`ray.train.get_all_reported_checkpoints`:

  * Always returns an empty list.
  * Historical checkpoint tracking isn't available.

* :func:`ray.train.collective.barrier`:

  * No-op in single-process mode.
  * In multi-process mode with torchrun, relies on PyTorch's distributed primitives.

* :func:`ray.train.collective.broadcast_from_rank_zero`:

  * Returns the data as-is in single-process mode without broadcasting.

>   * In multi-process mode with torchrun, relies on PyTorch's distributed primitives.

Reviewer 3: Do we need to document these? Is this not the same behavior?
